crates/shielded_token/src/masp/shielded_wallet.rs

```diff
@@ -713,95 +709,228 @@ pub trait ShieldedApi<U: ShieldedUtils + MaybeSend + MaybeSync>:
-                let mut est_conv = I128Sum::zero();
-                for ((_, asset_data), val) in decoded_conv.components() {
-                    let mut new_asset = asset_data.clone();
-                    if new_asset.epoch != Some(MaspEpoch::zero()) {
-                        new_asset.redate_to_next_epoch();
-                    }
-                    est_conv += ValueSum::from_pair(new_asset.encode()?, *val)
+                    asset_data.redate_to_next_epoch();
                 }
-                estimated_conversions.insert(
-                    asset.encode().unwrap(),
-                    (AllowedConversion::from(est_conv), wit.clone(), 0),
-                );
+
+                est_conv += ValueSum::from_pair(asset_data.encode()?, val)
             }
-            conversions.extend(estimated_conversions);
-            // use the estimations to convert the amount
-            let exchanged_amount = self
-                .compute_exchanged_amount(
-                    context.client(),
-                    context.io(),
-                    balance.clone(),
-                    target_epoch,
-                    conversions,
-                )
-                .await?
-                .0;
+            estimated_next_epoch_conversions.insert(
+                asset.encode().unwrap(),
+                (AllowedConversion::from(est_conv), wit.clone(), 0),
+            );
+        }
 
-            let rewards = exchanged_amount - balance;
-            // sum up the rewards.
-            Ok(self
-                .decode_sum(context.client(), rewards)
-                .await
-                .0
-                .components()
-                .filter(|((_, data), _)| {
-                    // this should always be true, but we check it anyway
-                    data.token == native_token
-                })
-                .map(|(_, val)| *val)
-                .sum::<i128>())
-        } else {
-            Ok(0)
+        // Get the estimated future balance based on the new conversions
+        let next_exchanged_balance = self
+            .compute_exchanged_amount(
+                context.client(),
+                context.io(),
+                current_exchanged_balance.clone(),
+                next_epoch,
+                estimated_next_epoch_conversions,
+            )
+            .await?
+            .0;
+
+        // Extract only the native token balance for rewards. Depending on
+        // rewards being enabled for the native token or not, we'll find the
+        // balance we look for at either epoch 0 or the current epoch. If
+        // rewards are not enabled for the native token we might have
+        // balances at any epochs in between these two but these would be
+        // guaranteed to be the same in both the exchanged amounts we've
+        // computed here above, meaning they are irrelevant for the
+        // estimation of the next epoch rewards since they would cancel out
+        // anyway
```

Comment:
```suggestion
        // Extract only the native token balance for rewards. Depending on
        // rewards being enabled for the native token or not, we'll find the
        // balance we look for at either epoch 0 or the current epoch. If
        // rewards are not enabled for the native token we might have
        // balances at any epochs in between these two. These would be
        // guaranteed to be the same in both the exchanged amounts we've
        // computed here above. This means they are irrelevant for the
        // estimation of the next epoch rewards since they would cancel out
        // anyway.
```

Reply:
nit: run on sentence